File: apps/libs/vending/softactivate/src/SoftActivate.Licensing/BitStream.cs
//
// Copyright (c) 2014 SoftActivate. All rights reserved.
//

using System;

namespace SoftActivate.Licensing
{
    class BitStream
    {
        public BitStream()
        {

        }

        public BitStream(int sizeInBits, bool randomize = false)
        {
            Create(sizeInBits, randomize);
        }

        public BitStream(byte[] buf, int bufBits)
        {
            Attach(buf, bufBits);
        }

        ~BitStream()
        {

        }

        public void Create(int bitCount, bool randomize = false)
        {
            buf = new byte[(bitCount + 7) >> 3];

            if (randomize)
            {
                Random rand = new Random();

                for (int i = 0; i < buf.Length; i++)
                    buf[i] = (byte)rand.Next(0, 255);
            }

            bufBits = bitCount;
            currentBitIndex = 0;
        }

        public void Attach(byte[] buf, int bufBits)
        {
            this.buf = buf;
            this.bufBits = bufBits;
            currentBitIndex = 0;
        }

        public int Read(byte[] readBuf, int bitCount)
        {
            if (bitCount + currentBitIndex > bufBits)
            {
                bitCount = bufBits - currentBitIndex;
            }

            if (bitCount == 0)
            {
                return 0;
            }

            int readBitCount = bitCount;
            byte lastByteMask = (byte)((1 << (readBitCount & 7)) - 1);
            int sourcePtr = (currentBitIndex >> 3);
            int byteCount = (bitCount + 7) >> 3;
            int destPtr = 0;

            int downShift = currentBitIndex & 0x7;
            int upShift = 8 - downShift;

            if (downShift == 0)
            {
                while (byteCount-- > 0)
                {
                    readBuf[destPtr++] = buf[sourcePtr++];
                }

                currentBitIndex += bitCount;

                if (lastByteMask > 0)
                    readBuf[readBitCount >> 3] &= lastByteMask;
                
                return readBitCount;
            }

            byte sourceByte = (byte)(buf[sourcePtr] >> downShift);
            currentBitIndex += bitCount;

            for (; bitCount > 8; bitCount -= 8)
            {
                byte nextByte = buf[++sourcePtr];
                readBuf[destPtr++] = (byte)(sourceByte | (nextByte << upShift));
                sourceByte = (byte)(nextByte >> downShift);
            }

            if (bitCount > 0)
            {
                if (bitCount <= upShift)
                {
                    readBuf[destPtr] = sourceByte;

                    if (lastByteMask > 0)
                        readBuf[readBitCount >> 3] &= lastByteMask;
                    
                    return readBitCount;
                }

                readBuf[destPtr] = (byte)(sourceByte | (buf[sourcePtr + 1] << upShift));
            }

            if (lastByteMask > 0)
                readBuf[readBitCount >> 3] &= lastByteMask;

            return readBitCount;
        }

        public int Write(byte[] writeBuf, int bitCount)
        {
            int maxBits = ((bufBits + 7) >> 3) << 3;

            if (bitCount + currentBitIndex > maxBits)
            {
                bitCount = maxBits - currentBitIndex;
            }

            if (bitCount <= 0)
                return 0;

            int writeBitCount = bitCount;
            int upShift = currentBitIndex & 0x7;
            int downShift = 8 - upShift;

            int sourcePtr = 0;
            int destPtr = currentBitIndex >> 3;

            // if this write is for <= 1 byte, and it will all fit in the
            // first dest byte, then do some special masking.
            if (downShift >= bitCount)
            {
                byte mask = (byte)(((1 << bitCount) - 1) << upShift);
                buf[destPtr] = (byte)((buf[destPtr] & ~mask) | ((writeBuf[sourcePtr] << upShift) & mask));
                currentBitIndex += bitCount;
                return writeBitCount;
            }

            // check for byte aligned writes -- this will be
            // much faster than the shifting writes.
            if (upShift == 0)
            {
                currentBitIndex += bitCount;

                for (; bitCount >= 8; bitCount -= 8)
                    buf[destPtr++] = writeBuf[sourcePtr++];

                if (bitCount > 0)
                {
                    byte mask = (byte)((1 << bitCount) - 1);
                    buf[destPtr] = (byte)((writeBuf[sourcePtr] & mask) | (buf[destPtr] & ~mask));
                }

                return writeBitCount;
            }

            // the write destination is not byte aligned.
            byte sourceByte;
            byte destByte = (byte)(buf[destPtr] & (0xFF >> downShift));
            byte lastMask = (byte)(0xFF >> (7 - ((currentBitIndex + bitCount - 1) & 0x7)));

            currentBitIndex += bitCount;

            for (; bitCount >= 8; bitCount -= 8)
            {
                sourceByte = writeBuf[sourcePtr++];
                buf[destPtr++] = (byte)(destByte | (sourceByte << upShift));
                destByte = (byte)(sourceByte >> downShift);
            }

            if (bitCount == 0)
            {
                if (destPtr < buf.Length) buf[destPtr] = (byte)((buf[destPtr] & ~lastMask) | (destByte & lastMask));
                return writeBitCount;
            }

            if (bitCount <= downShift)
            {
                if (destPtr < buf.Length) buf[destPtr] = (byte)((buf[destPtr] & ~lastMask) | ((destByte | (writeBuf[sourcePtr] << upShift)) & lastMask));
                return writeBitCount;
            }

            sourceByte = writeBuf[sourcePtr];

            buf[destPtr++] = (byte)(destByte | (sourceByte << upShift));
            buf[destPtr] = (byte)((buf[destPtr] & ~lastMask) | ((sourceByte >> downShift) & lastMask));

            return writeBitCount;
        }

        public int Seek(int offset)
        {
            return Seek(offset, false);
        }

        public int Seek(int offset, bool relative)
        {
            int oldIndex = currentBitIndex;
            currentBitIndex = (relative) ? currentBitIndex + offset : offset;
            return oldIndex;
        }

        public int Size
        {
            get
            {
                return bufBits;
            }
        }

        public void ZeroPadToNextByte()
        {
            byte[] zero = new byte[1] { 0 };

            if ((currentBitIndex & 0x7) > 0)
                Write(zero, 8 - (currentBitIndex & 0x7));
        }

        public byte[] Buffer
        {
            get
            {
                return buf;
            }
        }

        public void Clear()
        {
            for (int i = 0; i < (bufBits + 7) >> 3; i++)
            {
                buf[i] = 0;
            }
        }

        public void Randomize()
        {
            Random rand = new Random();
            rand.NextBytes(buf);
        }

        public int Length
        {
            get
            {
                return bufBits;
            }
        }

        protected byte[] buf;
        protected int bufBits;
        protected int currentBitIndex;
    }
}

File: apps/libs/vending/softactivate/src/Licensing/bitstream.cpp
//
// Copyright (c) 2014 FreeGrabApp. All rights reserved.
//

#include "precomp.h"

#include <stdio.h>
#include <string.h>

#ifdef _WIN32
#include <tchar.h>
#endif

#include "bitstream.h"

BitStream::BitStream()
{
    Init();
}

BitStream::BitStream(void * buf, int bufBits)
{
    Init();

    Attach(buf, bufBits);
}

BitStream::BitStream(int bitSize)
{
    Init();

    Attach(new unsigned char[(bitSize + 7) >> 3], bitSize);

    m_ownMemory = true;
}

BitStream::~BitStream()
{
    if (m_buf && m_ownMemory)
        delete [] m_buf;
}

void BitStream::Init()
{
    m_ownMemory = true;
    m_buf = NULL;
    m_bufBits = 0;
    m_currentBitIndex = 0;
}

bool BitStream::Create(int bitCount)
{
    m_ownMemory = true;
    m_bufBits = bitCount;
    m_currentBitIndex = 0;

    if (!bitCount)
    {
        m_ownMemory = false;
        return true;
    }

    int byteCount = (bitCount + 7) >> 3;

    if ((m_buf = new unsigned char[byteCount]) == NULL)
        return false;

    memset(m_buf, 0, byteCount);

    return true;
}

void BitStream::Attach(void * buf, int bufBits)
{
    if (m_buf && m_ownMemory)
        delete [] m_buf;

    m_buf = (unsigned char *)buf;
    m_bufBits = bufBits;
    m_currentBitIndex = 0;
    m_ownMemory = false;
}

int BitStream::Read(void *bitPtr, int bitCount)
{
    if(bitCount + m_currentBitIndex > m_bufBits)
    {
        bitCount = m_bufBits - m_currentBitIndex;
    }

    if(!bitCount)
    {
        return 0;
    }

    int readBitCount = bitCount;
    unsigned char lastByteMask = (unsigned char)((1 << (readBitCount & 7)) - 1);
    unsigned char * sourcePtr = m_buf + (m_currentBitIndex >> 3);
    int byteCount = (bitCount + 7) >> 3;
    unsigned char * destPtr = (unsigned char *)bitPtr;

    int downShift = m_currentBitIndex & 0x7;
    int upShift = 8 - downShift;

    if (!downShift)
    {
        while(byteCount--)
            *destPtr++ = *sourcePtr++;

        m_currentBitIndex += bitCount;

        if (lastByteMask > 0)
            ((unsigned char *)bitPtr)[readBitCount >> 3] &= lastByteMask;

        return readBitCount;
    }

    unsigned char sourceByte = *sourcePtr >> downShift;
    m_currentBitIndex += bitCount;

    for (; bitCount > 8; bitCount -= 8)
    {
        unsigned char nextByte = *++sourcePtr;
        *destPtr++ = sourceByte | (nextByte << upShift);
        sourceByte = nextByte >> downShift;
    }

    if (bitCount)
    {
        if (bitCount <= upShift)
        {
            *destPtr = sourceByte;
            return readBitCount;
        }

        *destPtr = sourceByte | ( (*++sourcePtr) << upShift);
    }

    if (lastByteMask > 0)
        ((unsigned char *)bitPtr)[readBitCount >> 3] &= lastByteMask;

    return readBitCount;
}

int BitStream::Write(const void * bitPtr, int bitCount)
{
   int maxBits = ((m_bufBits + 7) >> 3) << 3;

   if (bitCount + m_currentBitIndex > maxBits)
   {
        bitCount = maxBits - m_currentBitIndex;
   }

   if(!bitCount)
      return 0;

   int writeBitCount = bitCount;
   int upShift   = m_currentBitIndex & 0x7;
   int downShift = 8 - upShift;

   const unsigned char * sourcePtr = (unsigned char *)bitPtr;
   unsigned char * destPtr = m_buf + (m_currentBitIndex >> 3);

   // if this write is for <= 1 byte, and it will all fit in the
   // first dest byte, then do some special masking.
   if (downShift >= bitCount)
   {
      unsigned char mask = (unsigned char)(((1 << bitCount) - 1) << upShift);
      *destPtr = (*destPtr & ~mask) | ((*sourcePtr << upShift) & mask);
      m_currentBitIndex += bitCount;
      return writeBitCount;
   }

   // check for byte aligned writes -- this will be
   // much faster than the shifting writes.

   if(!upShift) 
   {
      m_currentBitIndex += bitCount;

      for(; bitCount >= 8; bitCount -= 8)
         *destPtr++ = *sourcePtr++;

      if(bitCount)
      {
         unsigned char mask = (unsigned char)((1 << bitCount) - 1);
         *destPtr = (*sourcePtr & mask) | (*destPtr & ~mask);
      }

      return writeBitCount;
   }

   // the write destination is not byte aligned.
   unsigned char sourceByte;
   unsigned char destByte = *destPtr & (0xFF >> downShift);
   unsigned char lastMask  = (unsigned char)(0xFF >> (7 - ((m_currentBitIndex + bitCount - 1) & 0x7)));

   m_currentBitIndex += bitCount;

   for(;bitCount >= 8; bitCount -= 8)
   {
      sourceByte = *sourcePtr++;
      *destPtr++ = destByte | (sourceByte << upShift);
      destByte = sourceByte >> downShift;
   }

   if(bitCount == 0)
   {
      *destPtr = (*destPtr & ~lastMask) | (destByte & lastMask);
      return writeBitCount;
   }

   if(bitCount <= downShift)
   {
      *destPtr = (*destPtr & ~lastMask) | ((destByte | (*sourcePtr << upShift)) & lastMask);
      return writeBitCount;
   }

   sourceByte = *sourcePtr;

   *destPtr++ = destByte | (sourceByte << upShift);
   *destPtr = (*destPtr & ~lastMask) | ((sourceByte >> downShift) & lastMask);

   return writeBitCount;
}

void BitStream::ZeroPadToNextByte()
{
    unsigned char zero = 0;

    if(m_currentBitIndex & 0x7)
        Write(&zero, 8 - (m_currentBitIndex & 0x7));
}

int BitStream::Seek(int offset, bool relative)
{
    unsigned oldIndex = m_currentBitIndex;
    m_currentBitIndex = (relative) ? m_currentBitIndex + offset : offset;
    return oldIndex;
}

int BitStream::GetSize()
{
    return m_bufBits;
}

void * BitStream::GetBuffer(int * sizeInBits)
{
    if (sizeInBits != NULL)
        *sizeInBits = m_bufBits;
    return m_buf;
}

void BitStream::ReleaseBuffer(int bitCount)
{
    m_currentBitIndex = bitCount;
}

void BitStream::Clear()
{
    memset(m_buf, 0, (m_bufBits + 7) >> 3);
}

